File: py/slack_sdk/features.py
# Slack SDK feature factory

from slack_sdk.feature.base_feature import SlackBaseFeature
from slack_sdk.feature.debug_feature import SlackDebugFeature
from slack_sdk.feature.idempotency_feature import SlackIdempotencyFeature
from slack_sdk.feature.metrics_feature import SlackMetricsFeature
from slack_sdk.feature.paging_feature import SlackPagingFeature
from slack_sdk.feature.ratelimit_feature import SlackRatelimitFeature
from slack_sdk.feature.retry_feature import SlackRetryFeature
from slack_sdk.feature.test_feature import SlackTestFeature
from slack_sdk.feature.timeout_feature import SlackTimeoutFeature


_FEATURES = {
    "base": lambda: SlackBaseFeature(),
    "debug": lambda: SlackDebugFeature(),
    "idempotency": lambda: SlackIdempotencyFeature(),
    "metrics": lambda: SlackMetricsFeature(),
    "paging": lambda: SlackPagingFeature(),
    "ratelimit": lambda: SlackRatelimitFeature(),
    "retry": lambda: SlackRetryFeature(),
    "test": lambda: SlackTestFeature(),
    "timeout": lambda: SlackTimeoutFeature(),
}


def _make_feature(name):
    factory = _FEATURES.get(name)
    if factory is not None:
        return factory()
    return _FEATURES["base"]()


# True when this SDK was generated with the named feature class - the
# constructor's tolerance for extend-carried features reads this (an
# active name with no generated class must not become a BaseFeature
# stray when an extend instance carries it).
def _has_feature(name):
    return name in _FEATURES

File: php/features.php
<?php
declare(strict_types=1);

// Slack SDK feature factory

require_once __DIR__ . '/feature/BaseFeature.php';
require_once __DIR__ . '/feature/DebugFeature.php';
require_once __DIR__ . '/feature/IdempotencyFeature.php';
require_once __DIR__ . '/feature/MetricsFeature.php';
require_once __DIR__ . '/feature/PagingFeature.php';
require_once __DIR__ . '/feature/RatelimitFeature.php';
require_once __DIR__ . '/feature/RetryFeature.php';
require_once __DIR__ . '/feature/TestFeature.php';
require_once __DIR__ . '/feature/TimeoutFeature.php';


class SlackFeatures
{
    public static function make_feature(string $name)
    {
        switch ($name) {
            case "base":
                return new SlackBaseFeature();
            case "debug":
                return new SlackDebugFeature();
            case "idempotency":
                return new SlackIdempotencyFeature();
            case "metrics":
                return new SlackMetricsFeature();
            case "paging":
                return new SlackPagingFeature();
            case "ratelimit":
                return new SlackRatelimitFeature();
            case "retry":
                return new SlackRetryFeature();
            case "test":
                return new SlackTestFeature();
            case "timeout":
                return new SlackTimeoutFeature();
            default:
                return new SlackBaseFeature();
        }
    }

    /**
     * Does a generated feature class back this name? False for a name only
     * an options extend instance can supply (the station adopt path) - the
     * constructor uses this to skip make_feature for such names instead of
     * adding a stray BaseFeature.
     */
    public static function has_feature(string $name): bool
    {
        switch ($name) {
            case "base":
            case "debug":
            case "idempotency":
            case "metrics":
            case "paging":
            case "ratelimit":
            case "retry":
            case "test":
            case "timeout":
                return true;
            default:
                return false;
        }
    }
}
